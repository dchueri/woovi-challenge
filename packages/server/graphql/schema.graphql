"""Root query"""
type Query {
  movies(
    """Returns the items in the list that come after the specified cursor."""
    after: String

    """Returns the first n items from the list."""
    first: Int

    """Returns the items in the list that come before the specified cursor."""
    before: String

    """Returns the last n items from the list."""
    last: Int
  ): MovieConnection!
  user(recovery: String!): User
}

"""A connection to a list of items."""
type MovieConnection {
  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [MovieEdge]
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!

  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String
}

"""An edge in a connection."""
type MovieEdge {
  """The item at the end of the edge"""
  node: Movie

  """A cursor for use in pagination"""
  cursor: String!
}

"""Movie Type"""
type Movie {
  id: String
  title: String
  genre: String
  image: String
  description: String
  average: Float
}

type User implements Node {
  """The ID of an object"""
  id: ID!
  name: String!
  email: String!
  recovery: String!
  helperSeen: Boolean
}

"""An object with an ID"""
interface Node {
  """The id of the object."""
  id: ID!
}

type Mutation {
  CreateMovie(input: CreateMovieInput!): CreateMoviePayload
  DeleteMovie(input: DeleteMovieInput!): DeleteMoviePayload
  UpdateMovie(input: UpdateMovieInput!): UpdateMoviePayload
  LoginMutation(input: UserLoginInput!): UserLoginPayload
  RegisterUserMutation(input: UserRegisterInput!): UserRegisterPayload
  UpdateUserMutation(input: UpdateUserInput!): UpdateUserPayload
  CommentCreateMutation(input: CommentCreateInput!): CommentCreatePayload
}

type CreateMoviePayload {
  movieEdge: MovieEdge
  error: String
  clientMutationId: String
}

input CreateMovieInput {
  title: String!
  genre: String!
  image: String!
  description: String!
  average: Float!
  clientMutationId: String
}

type DeleteMoviePayload {
  error: String
  deletedId: String
  clientMutationId: String
}

input DeleteMovieInput {
  id: String!
  clientMutationId: String
}

type UpdateMoviePayload {
  movieEdge: MovieEdge
  error: String
  clientMutationId: String
}

input UpdateMovieInput {
  id: String!
  title: String
  genre: String
  image: String
  description: String
  average: Float
  clientMutationId: String
}

type UserLoginPayload {
  token: String
  me: User
  error: String
  clientMutationId: String
}

input UserLoginInput {
  email: String!
  password: String!
  clientMutationId: String
}

type UserRegisterPayload {
  token: String
  me: User
  success: String
  error: String
  clientMutationId: String
}

input UserRegisterInput {
  name: String!
  email: String!
  password: String!
  clientMutationId: String
}

type UpdateUserPayload {
  success: String
  error: String
  clientMutationId: String
}

input UpdateUserInput {
  email: String!
  name: String
  password: String
  recovery: String
  helperSeen: Boolean
  clientMutationId: String
}

type CommentCreatePayload {
  commentEdge: CommentEdge
  movie: Movie
  error: String
  clientMutationId: String
}

"""An edge in a connection."""
type CommentEdge {
  """The item at the end of the edge."""
  node: Comment

  """A cursor for use in pagination."""
  cursor: String!
}

"""Comment data"""
type Comment implements Node {
  """The ID of an object"""
  id: ID!
  text: String
  user: User
  movie: Movie
  createdAt: String
  updatedAt: String
}

input CommentCreateInput {
  movie: ID!
  text: String!
  clientMutationId: String
}