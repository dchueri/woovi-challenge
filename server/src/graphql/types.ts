import mongoose from "mongoose";
import { IUser } from "../types/types";

import { DataLoaders } from "../graphql/loaderRegister";

declare type ObjectId = mongoose.Schema.Types.ObjectId;

export type GraphQLContext = {
  user?: IUser;
  dataloaders: DataLoaders;
};

export type LoaderFn = (
  ctx: GraphQLContext,
  id: string | ObjectId | object
) => any;
