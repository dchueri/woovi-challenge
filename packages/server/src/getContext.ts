import { Context, Request } from 'koa';

import { GraphQLContext } from './graphql/types';
import { getDataloaders } from './modules/loader/loaderRegister';
import { IUser } from './modules/user/UserModel';

type ContextVars = {
  user?: IUser | null;
  req?: Request;
  koaContext: Context;
  setCookie: (cookieName: string, token: string) => void;
};

export const getContext = async (ctx: any) => {
  const dataloaders = getDataloaders();

  return {
    req: ctx.req,
    dataloaders,
    user: ctx.user,
    koaContext: ctx.koaContext,
    setCookie: ctx.setCookie,
  } as GraphQLContext;
};
