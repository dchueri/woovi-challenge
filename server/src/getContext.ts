import { verify } from "jsonwebtoken";
import { Context, Request } from "koa";
import { getAllDataLoaders } from "./graphql/loaderRegister";
import { GraphQLContext } from "./graphql/types";
import UserModel from "./modules/user/UserModel";
import IUserContext from "./types/types";

type ContextVars = {
  user?: IUserContext;
  req: Request;
  context: Context;
};

export const getContext = async (ctx: ContextVars): Promise<GraphQLContext> => {
  const dataloaders = getAllDataLoaders();
  const authorization = ctx.req?.headers.authorization;
  if (!authorization) {
    return {
      req: ctx.req,
      dataloaders,
      user: null,
      context: ctx.context,
    } as GraphQLContext;
  }

  const token = authorization.split(" ");

  const authUser = verify(token[1], process.env.JWT_SECRET);
  if (!authUser) {
    return {
      req: ctx.req,
      dataloaders,
      user: null,
      context: ctx.context,
    } as GraphQLContext;
  }

  const user = await UserModel.findById(authUser.id);
  user.password = "***";
  return {
    req: ctx.req,
    dataloaders,
    user: user,
    context: ctx.context,
  } as GraphQLContext;
};
