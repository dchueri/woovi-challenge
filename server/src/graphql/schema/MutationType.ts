import { GraphQLObjectType } from "graphql";
import MovieMutations from "../../modules/movie/mutations";
import UserMutations from "../../modules/user/mutations";

const MutationType = new GraphQLObjectType({
  name: "Mutation",
  fields: () => ({
    ...MovieMutations,
    ...UserMutations,
  }),
});

export default MutationType;
