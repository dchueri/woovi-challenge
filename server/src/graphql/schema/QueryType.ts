import { GraphQLNonNull, GraphQLObjectType } from "graphql";
import { connectionArgs, connectionFromArray } from "graphql-relay";
import * as MovieLoader from "../../graphql/movie/MovieLoader";
import { MovieConnection } from "../../graphql/movie/MovieType";

const QueryType = new GraphQLObjectType({
  name: "Query",
  description: "Root query",
  fields: () => ({
    movies: {
      type: new GraphQLNonNull(MovieConnection),
      args: connectionArgs,
      resolve: async (_, args, context) => {
        const data = await MovieLoader.getAll();
        return connectionFromArray(data, args);
      },
    },
  }),
});

export default QueryType;
