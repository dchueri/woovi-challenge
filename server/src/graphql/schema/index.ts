import { GraphQLSchema } from "graphql";
import MutationType from "../../graphql/schema/MutationType";
import QueryType from "../../graphql/schema/QueryType";

export const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});
