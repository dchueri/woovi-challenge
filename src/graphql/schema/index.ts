import { GraphQLSchema } from "graphql";
import MutationType from "./MutationType";
import QueryType from "./QueryType";

export const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});
