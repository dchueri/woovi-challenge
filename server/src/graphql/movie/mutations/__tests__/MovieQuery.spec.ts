import { graphql } from "graphql";
import { createUser } from "../../../user/fixture/createUser";

import {
    clearDbAndRestartCounters,
    connectMongoose,
    disconnectMongoose,
    sanitizeTestObject
} from "../../../../../test";
import { getContext } from "../../../../getContext";
import { schema } from "../../../schema";
import { createMovie } from "../../fixture/createMovie";

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it("should be return all movies", async () => {
  const user = await createUser();

  await createMovie();

  // language=GraphQL
  const query = `
    query Q {
      movies {
        edges {
          node {
            title
          }
        }
      }
    }
  `;

  const rootValue = {};
  const contextValue = await getContext({ user });
  const variableValues = {};

  const result = await graphql({
    schema,
    source: query,
    rootValue,
    contextValue,
    variableValues,
  });

  expect(result.errors).toBeUndefined();
  expect(result.data.movies.edges.length).toBe(1);
  expect(sanitizeTestObject(result.data)).toMatchSnapshot();
});
