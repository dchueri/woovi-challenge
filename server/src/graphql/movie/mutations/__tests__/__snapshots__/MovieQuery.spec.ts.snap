// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be return all movies 1`] = `
{
  "movies": {
    "edges": [
      {
        "node": {
          "title": "content#0",
        },
      },
    ],
  },
}
`;

exports[`should be return an empty movies list 1`] = `
{
  "movies": {
    "edges": [],
  },
}
`;
